br/pkg/storage/gcs.go

```diff
@@ -447,39 +449,66 @@ skipHandleCred:
 	return ret, nil
 }
 
-// Reset resets the GCS storage.
+// Reset resets the GCS storage. Reset should not be used concurrently with
+// Close.
 func (s *GCSStorage) Reset(ctx context.Context) error {
 	logutil.Logger(ctx).Info("resetting gcs storage")
 
-	for _, client := range s.clients {
-		_ = client.Close()
-	}
+	s.cancelAndCloseGCSClients()
+
+	clientCtx, clientCancel := context.WithCancel(context.Background())
+	done := make(chan struct{})
+	defer close(done)
+	go func() {
+		select {
+		case <-done:
+		case <-ctx.Done():
+			clientCancel()
+		}
+	}()
 
 	s.clients = make([]*storage.Client, gcsClientCnt)
-	eg, egCtx := util.NewErrorGroupWithRecoverWithCtx(ctx)
+	firstErr := atomic.NewError(nil)
+	wg := util.WaitGroupWrapper{}
 	for i := range s.clients {
-		eg.Go(func() error {
-			client, err := storage.NewClient(egCtx, s.clientOps...)
+		wg.RunWithLog(func() {
+			client, err := storage.NewClient(clientCtx, s.clientOps...)
 			if err != nil {
-				return errors.Trace(err)
+				firstErr.CompareAndSwap(nil, err)
+				clientCancel()
+				return
 			}
 			client.SetRetry(storage.WithErrorFunc(shouldRetry), storage.WithPolicy(storage.RetryAlways))
 			s.clients[i] = client
-			return nil
 		})
 	}
```

Comment:
maybe this one
```go
	s.clients = make([]*storage.Client, 0, gcsClientCnt)
	wg := util.WaitGroupWrapper{}
	cliCh := make(chan *storage.Client)
	wg.RunWithLog(func() {
		for range gcsClientCnt {
			select {
			case cli := <-cliCh:
				s.clients = append(s.clients, cli)
			case <-ctx.Done():
				clientCancel()
				return
			}
		}
	})
	firstErr := atomic.NewError(nil)
	for range gcsClientCnt {
		wg.RunWithLog(func() {
			client, err := storage.NewClient(clientCtx, s.clientOps...)
			if err != nil {
				firstErr.CompareAndSwap(nil, err)
				clientCancel()
				return
			}
			client.SetRetry(storage.WithErrorFunc(shouldRetry), storage.WithPolicy(storage.RetryAlways))
			select {
			case cliCh <- client:
			case <-clientCtx.Done():
			}
		})
	}
```

Reply:
@D3Hunter I didn't get your point. Do you want to avoid the ugly code? ⬇️

```
	done := make(chan struct{})
	defer close(done)
	go func() {
		select {
		case <-done:
		case <-ctx.Done():
			clientCancel()
		}
	}()
```

Reply:
yes, the `done` channel is not needed, and I want the routine managed by waitGroup